File: engine/src/warden/file_error_watcher.py
"""Tail de arquivo + regex -> evento de erro, sem tocar no projeto observado.

Cobre rotação de log (inode trocou / arquivo truncado) e stacktrace multi-linha
(agrupa linhas até a próxima entrada começar, ou até ficar 2s em silêncio —
cobre o erro fatal que é a última coisa escrita antes do processo morrer).
"""

import re
import threading
import time
from collections.abc import Callable
from pathlib import Path

_ENTRY_START_RE = re.compile(r"^\[")
_POLL_INTERVAL = 1.0
_IDLE_FLUSH_AFTER = 2.0
_MAX_ENTRY_LINES = 200


def _is_new_entry_start(line: str) -> bool:
    return bool(_ENTRY_START_RE.match(line))


class FileErrorWatcher:
    def __init__(self, path: Path, patterns: list[str], on_error: Callable[[str], None]):
        self.path = path
        self._patterns = [re.compile(p) for p in patterns]
        self._on_error = on_error
        self._stop_event = threading.Event()
        self._thread: threading.Thread | None = None
        self._inode: int | None = None
        self._pos = 0
        self._pending: list[str] | None = None
        self._last_line_at = 0.0

    def start(self) -> None:
        self._thread = threading.Thread(target=self._run, daemon=True)
        self._thread.start()

    def stop(self) -> None:
        self._stop_event.set()
        if self._thread is not None:
            self._thread.join(timeout=_POLL_INTERVAL * 2)

    def _run(self) -> None:
        while not self._stop_event.is_set():
            self._poll_once()
            self._stop_event.wait(_POLL_INTERVAL)

    def _poll_once(self) -> None:
        try:
            stat = self.path.stat()
        except OSError:
            return  # arquivo ainda não existe / sumiu momentaneamente durante rotação

        if self._inode is None:
            self._inode = stat.st_ino
            self._pos = stat.st_size  # começa no fim — não reprocessa histórico
            return

        if stat.st_ino != self._inode or stat.st_size < self._pos:
            self._inode = stat.st_ino
            self._pos = 0

        if stat.st_size == self._pos:
            if self._pending is not None and time.time() - self._last_line_at > _IDLE_FLUSH_AFTER:
                self._flush_pending()
            return

        with self.path.open("r", errors="replace") as f:
            f.seek(self._pos)
            new_text = f.read()
            self._pos = f.tell()

        for line in new_text.splitlines():
            self._feed_line(line)
        self._last_line_at = time.time()

    def _feed_line(self, line: str) -> None:
        if self._pending is not None and not _is_new_entry_start(line):
            self._pending.append(line)
            if len(self._pending) > _MAX_ENTRY_LINES:
                self._flush_pending()
            return
        if self._pending is not None:
            self._flush_pending()
        self._pending = [line]

    def _flush_pending(self) -> None:
        assert self._pending is not None
        entry = "\n".join(self._pending)
        self._pending = None
        if any(pattern.search(entry) for pattern in self._patterns):
            self._on_error(entry)

File: agent/Warden.Domain/Watch/FileErrorWatcher.cs
using System.Text;
using System.Text.RegularExpressions;

namespace Warden.Domain.Watch;

/// <summary>
/// Tail de arquivo + regex -> callback de erro, sem tocar no projeto observado. Cobre rotação de log
/// e stacktrace multi-linha (agrupa linhas até a próxima entrada começar, ou até ficar em silêncio
/// por <see cref="IdleFlushAfter"/> — cobre o erro fatal que é a última coisa escrita antes do
/// processo morrer). Mirror de `file_error_watcher.py`.
/// </summary>
/// <remarks>
/// Python usa `stat.st_ino` pra distinguir "arquivo recriado com mesmo nome" (logrotate) de
/// "mesmo arquivo, só cresceu". .NET não expõe inode sem P/Invoke a uma struct `stat` cujo layout
/// varia por libc/arquitetura — em vez de arriscar isso, usamos <see cref="FileInfo.CreationTimeUtc"/>
/// como identidade do arquivo: no Linux (.NET usa `statx`/`STATX_BTIME` quando o filesystem suporta,
/// ext4/xfs/btrfs suportam) ela muda quando o arquivo é recriado, igual trocaria o inode. Truncamento
/// copytruncate (mesmo arquivo, size zerado) continua coberto pelo shrink-check abaixo, que não depende disso.
/// </remarks>
public sealed class FileErrorWatcher
{
    private static readonly TimeSpan PollInterval = TimeSpan.FromSeconds(1);
    private static readonly TimeSpan IdleFlushAfter = TimeSpan.FromSeconds(2);
    private const int MaxEntryLines = 200;

    private readonly string _path;
    private readonly List<Regex> _patterns;
    private readonly Action<string> _onError;
    private readonly ManualResetEventSlim _stopSignal = new(false);

    private Thread? _thread;
    private DateTime? _identity;
    private long _pos;
    private List<string>? _pending;
    private DateTime _lastLineAt;

    public FileErrorWatcher(string path, IEnumerable<string> patterns, Action<string> onError)
    {
        _path = path;
        _patterns = patterns.Select(p => new Regex(p)).ToList();
        _onError = onError;
    }

    public void Start()
    {
        _thread = new Thread(Run) { IsBackground = true };
        _thread.Start();
    }

    public void Stop()
    {
        _stopSignal.Set();
        _thread?.Join(TimeSpan.FromSeconds(PollInterval.TotalSeconds * 2));
    }

    private void Run()
    {
        while (!_stopSignal.IsSet)
        {
            PollOnce();
            _stopSignal.Wait(PollInterval);
        }
    }

    private void PollOnce()
    {
        FileInfo info;
        DateTime identity;
        try
        {
            info = new FileInfo(_path);
            if (!info.Exists) return; // arquivo ainda não existe / sumiu momentaneamente durante rotação
            identity = info.CreationTimeUtc;
        }
        catch (IOException)
        {
            return;
        }

        if (_identity is null)
        {
            _identity = identity;
            _pos = info.Length; // começa no fim — não reprocessa histórico
            return;
        }

        if (identity != _identity || info.Length < _pos)
        {
            _identity = identity;
            _pos = 0;
        }

        if (info.Length == _pos)
        {
            if (_pending is not null && DateTime.UtcNow - _lastLineAt > IdleFlushAfter)
            {
                FlushPending();
            }
            return;
        }

        var toRead = (int)(info.Length - _pos);
        var buffer = new byte[toRead];
        using (var stream = new FileStream(_path, FileMode.Open, FileAccess.Read, FileShare.ReadWrite))
        {
            stream.Seek(_pos, SeekOrigin.Begin);
            stream.ReadExactly(buffer);
        }
        _pos = info.Length;

        var newText = Utf8ReplaceFallback.GetString(buffer);
        var lines = newText.Split('\n');
        // texto terminado em \n sempre gera uma última linha "" no Split — não é conteúdo real
        // (mesma semântica de `str.splitlines()` do Python, que não conta essa linha vazia final).
        var lineCount = newText.EndsWith('\n') ? lines.Length - 1 : lines.Length;
        for (var i = 0; i < lineCount; i++)
        {
            FeedLine(lines[i].TrimEnd('\r'));
        }
        _lastLineAt = DateTime.UtcNow;
    }

    private void FeedLine(string line)
    {
        if (_pending is not null && !IsNewEntryStart(line))
        {
            _pending.Add(line);
            if (_pending.Count > MaxEntryLines) FlushPending();
            return;
        }
        if (_pending is not null) FlushPending();
        _pending = [line];
    }

    private void FlushPending()
    {
        var entry = string.Join("\n", _pending!);
        _pending = null;
        if (_patterns.Any(p => p.IsMatch(entry))) _onError(entry);
    }

    private static bool IsNewEntryStart(string line) => line.StartsWith('[');

    /// <summary>Mesmo comportamento de `errors="replace"` do Python — bytes inválidos viram U+FFFD em vez de derrubar o watcher.</summary>
    private static readonly Encoding Utf8ReplaceFallback =
        Encoding.GetEncoding("utf-8", EncoderFallback.ReplacementFallback, DecoderFallback.ReplacementFallback);
}
